### DISCUSSION TITLE
k3s/k3d authenticate to secure registry 401 UNAUTHORIZED

### DISCUSSION THREAD

**Discussion by the-scott-hand** [k3s-io/k3s#3830] [Q&A] [answered]
./k3d version
k3d version v4.4.6
k3s version v1.21.1-k3s (default)

Hi there, I am trying to configure my k3d cluster to pull images from a secure registry. I am running the command in the following way:

`k3d cluster create --servers 1 mycluster -p "8069:80@loadbalancer" --volume $K3D_PATH/registries.yaml:/etc/rancher/k3s/registries.yaml --volume $K3D_PATH/my.pem:/etc/ssl/certs/my.pem`

my registries.yaml is as follows

```
mirrors:
    my.registry.org:
        endpoint:
            - https://my.registry.org
configs:
    my.registry.org:
        tls:
            ca_file: "/etc/ssl/certs/my.pem"
```

without the ca file I get x509 certificate signed by unknown authority, so I assume that my registries.yaml is being used in some way. However, there seems to be no combination of ca_file cert_file and key_file that get me past the `failed to pull and unpack image <image> failed to resolve reference: <image>  "failed to authorize: failed to fetch anonymous token: unexpected status: 401 UNAUTHORIZED. `

When i copy my.pem to /etc/docker/certs.d/my.registry.org/ as client.key and client.cert, I am able to docker pull as expected without logging in. however specifying these in the configs section and mounting them does not seem to enable k3s to pull images from the secure part. 

Where am i going wrong? Can anyone help me configure this correctly?

**Comment by brandond** [comment]
Your registry appears to require authentication, but you're not providing any? Did you want to provide a username and password, or cert_file and key_file for certificate auth?

**Reply by brandond** [reply]
I'm not sure what's necessary to get these mounted into k3d, you might ask this same question over in the k3d repo.

**Reply by the-scott-hand** [reply]
To clarify, I am using a secure registry and authenticate using registered server certs rather than a username/password situation. When using docker, copying and pasting my cert into files client.key and client.cert and placing them in /etc/docker/certs.d/my.registry.org/ is enough to get docker to pull from the secure registry. How can I configure my registries.yaml to accomplish this in k3s? I have experimented with using cert_file and key_file in the registries.yaml with this pem but I get the same error.

As far as mounting them into k3d, that is accomplished via the `--volume $K3D_PATH/my.pem:/etc/ssl/certs/my.pem` flag which I have confirmed I can do successfully, I didnt intend to ask any questions about mounting the certs. 

Thanks for the quick response!

**Reply by brandond** [reply]
Well, the docs are at https://rancher.com/docs/k3s/latest/en/installation/private-registry/#configs - if cert_file and key_file don't work then I'm not sure what to suggest. This is all just passed through directly to containerd, which does the actual image pulling.

Can you get this to work properly with k3s? If so then it's probably a k3d issue and not a problem with how you're setting up registries.yaml

**Reply by the-scott-hand** [reply]
Hmmm ok I will try asking in the k3d git discussions, I wasnt sure which place to ask. Thanks!

**Comment by ChristianCiach** [comment]
According to your question over at https://github.com/rancher/k3d/discussions/705, it looks like you're trying to use the **same** file for all three attributes: `ca_file`, `key_file` and `cert_file`. I wouldn't be surprised if containerd just doesn't support this.

Try to open your `my.pem` with a text editor and split this file into three separate files.

**Reply by brandond** [reply]
Yeah, that definitely wouldn't work...

**Reply by the-scott-hand** [reply]
not sure if i should keep responding here or in https://github.com/rancher/k3d/discussions/705 but I am going to separate my pem into the two files and go back to using my ca cert and give it a shot and let you guys know. thanks again

**Reply by ChristianCiach** [reply]
> not sure if i should keep responding here or in rancher/k3d#705

Seeing that you are using a bind mount to inject your `registries.yaml` (instead of using `k3d cluster create --registry-config`, which isn't an option in your case) I think this is the more appropriate project. *If* there is a bug involved, I guess it will be closer to K3s than to K3d.

**Reply by brandond** [reply]
Just to be clear, you need:
* one file for the CA certificate used to sign the registry's server cert (ca_file)
* one file for the k3s client's certificate, which is trusted by the registry (cert_file)
* one file for the k3s client certificate's private key (key_file)

**Reply by the-scott-hand** [reply]
I have separated the files and mounted them individually, but I still get the same error 401 unauthorized. 
```
mirrors:
    my.registry.org:
        endpoint:
            - https://my.registry.org
configs:
    my.registry.org:
        tls:
            ca_file: "/etc/ssl/certs/my.ca"
            key_file: "/etc/ssl/certs/my.key"
            cert_file: "/etc/ssl/certs/my.cert"
```
again these files do work for my docker install, so im not sure what im doing wrong. Is there somewhere I can get a more detailed error message? I tried exec-ing into the docker container for server-0 and running crictl pull <image> but i get the same error and --debug doesnt add anything extra. I keep getting
`pulling image: rpc error: code = Unknown desc = failed to pull and unpack image "my.registry.org/my-project/my-image:1.0.0": failed to resolve reference "my.registry.org/my-project/my-image:1.0.0": failed to authorize: failed to fetch anonymous token: unexpected status: 401 UNAUTHORIZED`

**Reply by the-scott-hand** [reply]
Additional interesting point of data for me.

If I run: `k3d cluster create --servers 1 mycluster -p "8069:80@loadbalancer"`
then I get x509 certificate signed by unknown authority

But if I run: `k3d cluster create --servers 1 mycluster -p "8069:80@loadbalancer" --volume $K3D_PATH/my.ca:/etc/ssl/certs/my.ca`
without any registries yaml i get 401 UNAUTHORIZED

**Reply by brandond** [reply]
Are you trying to use the registry's server cert and key as the K3s client cert? Or did you generate a new cert and key that are also signed by the registry's server CA?

**Reply by ChristianCiach** [reply]
I would love to help you out, but I don't really have any more ideas based on the information that you've provided.

Maybe you could check this: Inside your Container running a k3s server or agent, what does `/var/lib/rancher/k3s/agent/etc/containerd/config.toml` look like? Does the file include references to both certificates and the private key you've stated inside your `registries.yaml`?

If the containerd-config looks correct, maybe you can try to use Curl (from inside your container) to connect to your registry:
```
curl -v -k --key /etc/ssl/certs/my.key --cert /etc/ssl/certs/my.cert --cacert /etc/ssl/certs/my.ca https://my.registry.org/v2/
```
Does this work or do you get a 401 response, too?

**Reply by the-scott-hand** [reply]
A new cert and key that is specific to my vm. 

from inside the server-0 container i can run the following and successfully pull the image
`./bin/ctr images pull ---tlscacert /etc/ssl/certs/client.ca --tlscert /etc/ssl/certs/client.cert --tlskey /etc/ssl/certs/client.key my.registry.org/my-project/my-image:1.0.0`

but using those same files in my registries yaml does not work. any ideas?

**Reply by the-scott-hand** [reply]
and yes my toml does include all the references to the tls config. I was able to try adding the auth section with username and password and that worked, but I was hoping to not use a user/password combo in this file.

**Reply by the-scott-hand** [reply]
multiple users with sudo access on the same server with different docker accounts, trying to avoid putting their passwords in plaintext. wanted to run multiple instances of k3d. The server cert registered with the registry was working well for us with docker to solve this issue, and it seems ctr is able to pull the images when given those flags from inside the k3d container but just not k3s?

**Reply by the-scott-hand** [reply]
any ideas on how i can debug why ctr can pull my image from inside the container but kubernetes cant?

**Reply by iwilltry42** [reply]
@the-scott-hand , so if you add the files in the `registries.yaml`, they show up correctly in the containerd config `toml`?
Does `ctr image pull` **without** the tlscacert/tlscert/tlskey flags work then or not?

**Reply by the-scott-hand** [reply]
the ctr image pull without the flags does not work. Im not super familiar with the config.toml but it populated with values from my registries yaml

**Comment by the-scott-hand** [comment, marked as answer]
Not sure which of my many changes fixed the issue, but i got it working with the certs. These are some differences:

I switched to using the --config flag and specified all of my volumes and ports in that config file
**My mounts previously had nodeFilter `server[0]` where I now have `server[*]` and `agent[*]` (configured for 1 server no agents)**
The loadbalancer is currently disabled and wasnt previously
traefik is also disabled and wasnt previously
I have mounted my hosts /etc/resolv.conf onto /etc/resolv.conf in k3d
No longer mounting a "registries.yaml" just specified in the config file

Thanks for the responses

**Reply by brandond** [reply]
Yeah, you would probably want the registries configuration on all the servers and agents right? They all pull images...

**Reply by the-scott-hand** [reply]
yeah definitely, I just only had 1 server and no agents so i assumed server[0] was the full list
